src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs

```diff
@@ -1243,39 +1250,89 @@ private void EmitSwitchDispatch(BoundSwitchDispatch dispatch)
                 }
             }
 
-            Cci.IReference stringEqualityMethodRef = _module.Translate(equalityMethod, syntaxNode, _diagnostics);
+            Cci.IMethodReference stringEqualityMethodRef = null;
+            Cci.IMethodReference lengthRef = null;
 
-            Cci.IMethodReference stringLengthRef = null;
-            var stringLengthMethod = _module.Compilation.GetSpecialTypeMember(SpecialMember.System_String__Length) as MethodSymbol;
-            if (stringLengthMethod != null && !stringLengthMethod.HasUseSiteError)
+            Cci.IMethodReference sequenceEqualsRef = null;
+            Cci.IMethodReference asSpanRef = null;
+
+            if (isSpanOrReadOnlySpan)
             {
-                stringLengthRef = _module.Translate(stringLengthMethod, syntaxNode, _diagnostics);
+                var sequenceEqualsTMethod = _module.Compilation.GetWellKnownTypeMember(isReadOnlySpan
+                    ? WellKnownMember.System_MemoryExtensions__SequenceEqual_ReadOnlySpan_T
+                    : WellKnownMember.System_MemoryExtensions__SequenceEqual_Span_T) as MethodSymbol;
+                var sequenceEqualsCharMethod = sequenceEqualsTMethod?.Construct(_module.Compilation.GetSpecialType(SpecialType.System_Char));
+                Debug.Assert(sequenceEqualsCharMethod != null && !sequenceEqualsCharMethod.HasUseSiteError);
+                sequenceEqualsRef = _module.Translate(sequenceEqualsCharMethod, null, _diagnostics);
+
+                var asSpanMethod = _module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_MemoryExtensions__AsSpanString) as MethodSymbol;
+                Debug.Assert(asSpanMethod != null && !asSpanMethod.HasUseSiteError);
+                asSpanRef = _module.Translate(asSpanMethod, null, _diagnostics);
+
+                var spanTLengthMethod = _module.Compilation.GetWellKnownTypeMember(isReadOnlySpan
+                    ? WellKnownMember.System_ReadOnlySpan_T__get_Length
+                    : WellKnownMember.System_Span_T__get_Length) as MethodSymbol;
+                var spanCharLengthMethod = spanTLengthMethod?.AsMember((NamedTypeSymbol)keyType);
+                if (spanCharLengthMethod != null && !spanCharLengthMethod.HasUseSiteError)
```

Comment:
>spanCharLengthMethod != null && !spanCharLengthMethod.HasUseSiteError [](http://example.com/codeflow?start=20&length=69)

It seems surprising that we allow `Span<T>.Length` to be missing. (It looks like we allow missing `String.Length` for the existing string switch cases, and emit slightly different code, although I'm not sure why.) I think it would be better to require `Span<T>.Length` and `ReadOnlySpan<T>.Length` in `Binder` where we check for other well-known members needed for the span cases, and remove the conditional code here.